torch/_lobpcg.py

```diff
@@ -0,0 +1,719 @@
+          space between matrices), here we introduce element-wise
+          products (denoted by symbol `*`) so that the Rayleight-Ritz
+          procedure becomes::
+
+            SAS = S^T A S
+            SBS = S^T B S
+            d = (<diagonal of SBS>) ** -1/2    # this is 1-d column vector
+            dd = d d^T                         # this is 2-d matrix
+            R^T R = Cholesky(dd * SBS)
+            Ri = R^-1 * d                      # broadcasting
+            solve symeig problem Ri^T SAS Ri Z = Theta Z
+            C = Ri Z
+
+          where `dd` is 2-d matrix that replaces matrix products `D M
+          D` with one element-wise product `M * dd`; and `d` replaces
+          matrix product `D M` with element-wise product `M *
+          d`. Also, creating the diagonal matrix `D` is avoided.
+
+        Arguments:
+        S (Tensor): the matrix basis for the search subspace, size is
+                    :math:`(m, n)`.
+
+        Returns:
+        Ri (tensor): upper-triangular transformation matrix of size
+                     :math:`(n, n)`.
+
+        """
+        B = self.B
+        mm = torch.matmul
+        SBS = _utils.qform(B, S)
+
+        d1 = SBS.diagonal(0, -2, -1) ** -0.5
+        if d1.is_cuda:
+            # Avoid `test_torch.TestTorchDeviceTypeCUDA.test_lobpcg_basic_cuda_float64 leaked 4096 bytes CUDA memory on device 0`
+            d = torch.zeros((d1.shape[0], 1), dtype=d1.dtype, device=d1.device)
+            d[:, 0] = d1
+        else:
+            d = d1.reshape(d1.shape[0], 1)
+
+        dd = mm(d, _utils.transpose(d))
+        R = torch.cholesky(dd * SBS, upper=True)
+        # TODO: could use LAPACK ?trtri as R is upper-triangular
+        Rinv = torch.inverse(R)
+        return Rinv * d
+
+    def _get_svqb(self,
+                  U,     # Tensor
+                  drop,  # bool
+                  tau    # float
+                  ):
+        """Return B-orthonormal U.
+
+        .. note:: When `drop` is `False` then `svqb` is based on the
+                  Algorithm 4 from [DuerschPhD2015] that is a slight
+                  modification of the corresponding algorithm
+                  introduced in [StathopolousWu2002].
+
+        Arguments:
+
+          U (Tensor) : initial approximation, size is (m, n)
+          drop (bool) : when True, drop columns that
+                     contribution to the `span([U])` is small.
+          tau (float) : positive tolerance
+
+        Returns:
+
+          U (Tensor) : B-orthonormal columns (:math:`U^T B U = I`), size
+                       is (m, n1), where `n1 = n` if `drop` is `False,
+                       otherwise `n1 <= n`.
+
+        """
+        # type: (Tensor, bool, float) -> Tensor
+        if torch.numel(U) == 0:
+            return U
+        UBU = _utils.qform(self.B, U)
+        d = UBU.diagonal(0, -2, -1)
+        # detect and drop zero columns from U
+        nz = torch.where(abs(d) != 0.0)
```

Comment:
The condition here is never gonna be satisfied. The zero columns are better detected by their norm being smaller than some tolerance.

Reply:
I appreciate the use of "almost surely" adverb as your first statement holds for random data.
However, it is possible to construct input to `lobpcg` where `abs(d) == 0.0` actually holds and that will break the original algorithm (notice the `d ** -0.5` operations) from the method paper where the authors also presumed random input data.
So, this block of code is an extension of the algorithm to the extreme (but simple) cases that the authors did not consider. As a resolution of this conversation, I'll make the following comment in the code:
```
        # Detect and drop exact zero columns from U. While the test
        # `abs(d) == 0` is unlikely to be True for random data, it is
        # possible to construct input data to lobpcg where it will be
        # True leading to a failure (notice the `d ** -0.5` operation
        # in the original algorithm). To prevent the failure, we drop
        # the exact zero columns here and then continue with the
        # original algorithm below.
```

Reply:
But what if just replacing with `abs(d) < epsilon`?

Reply:
Your suggestion will influence the behavior of the original algorithm.
Without hard evidence of the actual gain or apparent bug, I would not modify the original algorithm that has been published in a peer-reviewed journal and has been tested by the authors.

Reply:
pls have a look at how this is done in
https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/eigen/lobpcg/lobpcg.py
you may want to be consistent to simplify the future maintenance.